olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java

```diff
@@ -263,6 +279,362 @@ public int hashCode() {
+                return false;
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "DerefedProperty(" +
+                    "isDerefed=" + isDerefed +
+                    ", innerConf=" + innerConf +
+                    ", innerValue='" + innerValue + '\'' +
+                    ", propName='" + propName + '\'' +
+                    ", locationContext='" + locationContext + '\'' +
+                    ')';
+        }
+    }
+
+    /**
+     * See {@link #structuralEquals(List, List, String, String)} for description of behavior.
+     */
+    private static boolean innerStructuralEquals(Map<String, ConfigurationData> a, Map<String, ConfigurationData> b, String aName, String bName) {
+        Optional<StructuralConfigurationData> aRootOpt = StructuralConfigurationData.fromName(a, aName);
+        Optional<StructuralConfigurationData> bRootOpt = StructuralConfigurationData.fromName(b, bName);
+        if (aRootOpt.isPresent() && bRootOpt.isPresent()) {
+            return aRootOpt.get().equals(bRootOpt.get());
+        } else {
+            if (aRootOpt.isPresent()) {
+                logger.fine(String.format("Component with name: %s not found in b", bName));
+            } else {
+                logger.fine(String.format("Component with name: %s not found in a", aName));
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Checks whether two ConfigurationData objects are 'structurally equal'. Two objects are structurally
+     * equal when their classNames are the same and when all of their properties are equal or, if those
+     * properties refer to another ConfigurationData by name, if all of their properties are structurally
+     * equal recursively.
+     *
+     * <p>
+     *
+     * N.B. Because the serialized configuration format internally represents things as strings with no type information
+     * but equality semantics often differ between strings and the types the represent, the method attempts to intuit
+     * represented types by attempting to parse the string as that type and treating it as that type if it successfully
+     * parses. This means that, eg. ID values that are typed as strings but are fully numerical will be converted to
+     * doubles and be compared for equality that way. For more details on the processing see {@link DerefedProperty#equals(Object)}.
+     *
+     * <p>
+     *
+     * {@code aName} should be the name of an element of {@code a} that is to be compared to {@code bName}
+     * in {@code b}. {@code a} and {@code b} should each contain all the ConfigurationData objects
+     * needed to instantiate the objects named by {@code aName} and {@code bName} respectively. Objects not
+     * instantiated by traversing children of {@code aName} and {@code bName} are ignored.
+     *
+     * <p>
+     *
+     * At log-level {@code FINE} this reports the first place where the two instances differ, and the nature of their
```

Comment:
Maybe expand this slightly? It logs all the places where the two instances differ in whatever object they differ in. Possibly `reports the first object where the two instances differ`, but that's not particularly clear either.